### PR TITLE
Dispatch concatenate where possible.

### LINKED ISSUE #4662: Try np.concatenate() for unregistered backends.
As discussed in https://github.com/plures/gumath/issues/22, unregistered backends work if `np.concatenate()` is tried.  This is a an example that deliberately does not touch the code path for registered backends but works for `xnd.array`:

```
diff --git a/dask/array/core.py b/dask/array/core.py
index f04ff0c..dd968a6 100644
--- a/dask/array/core.py
+++ b/dask/array/core.py
@@ -3505,6 +3505,13 @@ def concatenate3(arrays):
         x = unpack_singleton(arrays)
         return _concatenate2(arrays, axes=list(range(x.ndim)))
 
+    if hasattr(advanced, '__array_function__'):
+        try:
+            x = unpack_singleton(arrays)
+            return _concatenate2(arrays, axes=list(range(x.ndim)))
+        except TypeError:
+            pass # np.concatenate not implemented.
+
     ndim = ndimlist(arrays)
     if not ndim:
         return arrays
```

### PR DESCRIPTION
**Pull request by hameerabbasi**
- [x] Tests added / passed
- [x] Passes `flake8 dask`

I'm not sure what's causing the failures, it seems to be something about the call stack. If someone could help me figure it out, that'd be great.

Closes #4662

### CODE REVIEW

**Comment by jrbourbeau** [comment]
> I'm not sure what's causing the failures, it seems to be something about the call stack.

What failures are you getting locally?

**Comment by hameerabbasi** [comment]
Ah, I expected CI to show them too. But I also don't see why it would given that it happened without my changes too.

Here's the failure summary:

```
=================================== FAILURES ===================================
_________________________________ test_persist _________________________________

    def test_func():
        del _global_workers[:]
        _global_clients.clear()
        active_threads_start = set(threading._active)

        reset_config()

        dask.config.set({'distributed.comm.timeouts.connect': '5s'})
        # Restore default logging levels
        # XXX use pytest hooks/fixtures instead?
        for name, level in logging_levels.items():
            logging.getLogger(name).setLevel(level)

        result = None
        workers = []

        with pristine_loop() as loop:
            with check_active_rpc(loop, active_rpc_timeout):
                @gen.coroutine
                def coro():
                    with dask.config.set(config):
                        s = False
                        for i in range(5):
                            try:
                                s, ws = yield start_cluster(
                                    ncores, scheduler, loop, security=security,
                                    Worker=Worker, scheduler_kwargs=scheduler_kwargs,
                                    worker_kwargs=worker_kwargs)
                            except Exception as e:
                                logger.error("Failed to start gen_cluster, retrying", exc_info=True)
                            else:
                                workers[:] = ws
                                args = [s] + workers
                                break
                        if s is False:
                            raise Exception("Could not start cluster")
                        if client:
                            c = yield Client(s.address, loop=loop, security=security,
                                             asynchronous=True, **client_kwargs)
                            args = [c] + args
                        try:
                            future = func(*args)
                            if timeout:
                                future = gen.with_timeout(timedelta(seconds=timeout),
                                                          future)
                            result = yield future
                            if s.validate:
                                s.validate_state()
                        finally:
                            if client:
                                yield c._close(fast=s.status == 'closed')
                            yield end_cluster(s, workers)
                            yield gen.with_timeout(timedelta(seconds=1),
                                                   cleanup_global_workers())

                        try:
                            c = yield default_client()
                        except ValueError:
                            pass
                        else:
                            yield c._close(fast=True)

                        raise gen.Return(result)

                result = loop.run_sync(coro, timeout=timeout * 2 if timeout else timeout)

            for w in workers:
                if getattr(w, 'data', None):
                    try:
                        w.data.clear()
                    except EnvironmentError:
                        # zict backends can fail if their storage directory
                        # was already removed
                        pass
                    del w.data
            DequeHandler.clear_all_instances()
            for w in _global_workers:
                w = w()
                w._close(report=False, executor_wait=False)
                if w.status == 'running':
                    w.close()
            del _global_workers[:]

        if PY3 and not WINDOWS and check_new_threads:
            start = time()
            while True:
                bad = [t for t, v in threading._active.items()
                       if t not in active_threads_start and
                      "Threaded" not in v.name and
                      "watch message" not in v.name and
                      "TCP-Executor" not in v.name]
                if not bad:
                    break
                else:
                    sleep(0.01)
                if time() > start + 5:
                    from distributed import profile
                    tid = bad[0]
                    thread = threading._active[tid]
                    call_stacks = profile.call_stack(sys._current_frames()[tid])
>                   assert False, (thread, call_stacks)
E                   AssertionError: (<Thread(Profile, started daemon 123146351648768)>, ['  File "/Users/hameerabbasi/anaconda/envs/dask/lib/python3.7/thr...asi/anaconda/envs/dask/lib/python3.7/site-packages/distributed/profile.py", line 251, in _watch
E                     \tsleep(interval)
E                     '])
E                   assert False

../../anaconda/envs/dask/lib/python3.7/site-packages/distributed/utils_test.py:945: AssertionError
----------------------------- Captured stderr call -----------------------------
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:51116
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:51117
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:51117
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:51116
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                   34.36 GB
distributed.worker - INFO -       Local Directory: /Users/hameerabbasi/Quansight/dask/dask-worker-space/worker-1gs25rhd
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:51118
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:51118
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:51116
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          2
distributed.worker - INFO -                Memory:                   34.36 GB
distributed.worker - INFO -       Local Directory: /Users/hameerabbasi/Quansight/dask/dask-worker-space/worker-v21491p2
distributed.worker - INFO - -------------------------------------------------
distributed.scheduler - INFO - Register tcp://127.0.0.1:51117
distributed.scheduler - INFO - Register tcp://127.0.0.1:51118
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:51117
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:51118
distributed.core - INFO - Starting established connection
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:51116
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:51116
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Receive client connection: Client-827ab902-5854-11e9-b2a9-e0d55eb2634b
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Remove client Client-827ab902-5854-11e9-b2a9-e0d55eb2634b
distributed.scheduler - INFO - Remove client Client-827ab902-5854-11e9-b2a9-e0d55eb2634b
distributed.scheduler - INFO - Close client connection: Client-827ab902-5854-11e9-b2a9-e0d55eb2634b
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:51117
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:51118
distributed.scheduler - INFO - Scheduler closing...
distributed.scheduler - INFO - Scheduler closing all comms
distributed.scheduler - INFO - Remove worker tcp://127.0.0.1:51117
distributed.core - INFO - Removing comms to tcp://127.0.0.1:51117
distributed.scheduler - INFO - Remove worker tcp://127.0.0.1:51118
distributed.core - INFO - Removing comms to tcp://127.0.0.1:51118
distributed.scheduler - INFO - Lost all workers
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - -------------------------------------------------
____________________ test_local_get_with_distributed_active ____________________

    def test_func():
        del _global_workers[:]
        _global_clients.clear()
        active_threads_start = set(threading._active)

        reset_config()

        dask.config.set({'distributed.comm.timeouts.connect': '5s'})
        # Restore default logging levels
        # XXX use pytest hooks/fixtures instead?
        for name, level in logging_levels.items():
            logging.getLogger(name).setLevel(level)

        result = None
        workers = []

        with pristine_loop() as loop:
            with check_active_rpc(loop, active_rpc_timeout):
                @gen.coroutine
                def coro():
                    with dask.config.set(config):
                        s = False
                        for i in range(5):
                            try:
                                s, ws = yield start_cluster(
                                    ncores, scheduler, loop, security=security,
                                    Worker=Worker, scheduler_kwargs=scheduler_kwargs,
                                    worker_kwargs=worker_kwargs)
                            except Exception as e:
                                logger.error("Failed to start gen_cluster, retrying", exc_info=True)
                            else:
                                workers[:] = ws
                                args = [s] + workers
                                break
                        if s is False:
                            raise Exception("Could not start cluster")
                        if client:
                            c = yield Client(s.address, loop=loop, security=security,
                                             asynchronous=True, **client_kwargs)
                            args = [c] + args
                        try:
                            future = func(*args)
                            if timeout:
                                future = gen.with_timeout(timedelta(seconds=timeout),
                                                          future)
                            result = yield future
                            if s.validate:
                                s.validate_state()
                        finally:
                            if client:
                                yield c._close(fast=s.status == 'closed')
                            yield end_cluster(s, workers)
                            yield gen.with_timeout(timedelta(seconds=1),
                                                   cleanup_global_workers())

                        try:
                            c = yield default_client()
                        except ValueError:
                            pass
                        else:
                            yield c._close(fast=True)

                        raise gen.Return(result)

                result = loop.run_sync(coro, timeout=timeout * 2 if timeout else timeout)

            for w in workers:
                if getattr(w, 'data', None):
                    try:
                        w.data.clear()
                    except EnvironmentError:
                        # zict backends can fail if their storage directory
                        # was already removed
                        pass
                    del w.data
            DequeHandler.clear_all_instances()
            for w in _global_workers:
                w = w()
                w._close(report=False, executor_wait=False)
                if w.status == 'running':
                    w.close()
            del _global_workers[:]

        if PY3 and not WINDOWS and check_new_threads:
            start = time()
            while True:
                bad = [t for t, v in threading._active.items()
                       if t not in active_threads_start and
                      "Threaded" not in v.name and
                      "watch message" not in v.name and
                      "TCP-Executor" not in v.name]
                if not bad:
                    break
                else:
                    sleep(0.01)
                if time() > start + 5:
                    from distributed import profile
                    tid = bad[0]
                    thread = threading._active[tid]
                    call_stacks = profile.call_stack(sys._current_frames()[tid])
>                   assert False, (thread, call_stacks)
E                   AssertionError: (<Thread(Profile, started daemon 123146362159104)>, ['  File "/Users/hameerabbasi/anaconda/envs/dask/lib/python3.7/thr...asi/anaconda/envs/dask/lib/python3.7/site-packages/distributed/profile.py", line 251, in _watch
E                     \tsleep(interval)
E                     '])
E                   assert False

../../anaconda/envs/dask/lib/python3.7/site-packages/distributed/utils_test.py:945: AssertionError
----------------------------- Captured stderr call -----------------------------
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:51123
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:51124
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:51124
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:51123
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                   34.36 GB
distributed.worker - INFO -       Local Directory: /Users/hameerabbasi/Quansight/dask/dask-worker-space/worker-e5yhu04d
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:51125
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:51125
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:51123
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          2
distributed.worker - INFO -                Memory:                   34.36 GB
distributed.worker - INFO -       Local Directory: /Users/hameerabbasi/Quansight/dask/dask-worker-space/worker-ygm5mnhy
distributed.worker - INFO - -------------------------------------------------
distributed.scheduler - INFO - Register tcp://127.0.0.1:51124
distributed.scheduler - INFO - Register tcp://127.0.0.1:51125
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:51124
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:51125
distributed.core - INFO - Starting established connection
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:51123
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:51123
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Receive client connection: Client-8592f35c-5854-11e9-b2a9-e0d55eb2634b
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Remove client Client-8592f35c-5854-11e9-b2a9-e0d55eb2634b
distributed.scheduler - INFO - Remove client Client-8592f35c-5854-11e9-b2a9-e0d55eb2634b
distributed.scheduler - INFO - Close client connection: Client-8592f35c-5854-11e9-b2a9-e0d55eb2634b
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:51124
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:51125
distributed.scheduler - INFO - Scheduler closing...
distributed.scheduler - INFO - Scheduler closing all comms
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - -------------------------------------------------
________________________ test_serializable_groupby_agg _________________________

    def test_func():
        del _global_workers[:]
        _global_clients.clear()
        active_threads_start = set(threading._active)

        reset_config()

        dask.config.set({'distributed.comm.timeouts.connect': '5s'})
        # Restore default logging levels
        # XXX use pytest hooks/fixtures instead?
        for name, level in logging_levels.items():
            logging.getLogger(name).setLevel(level)

        result = None
        workers = []

        with pristine_loop() as loop:
            with check_active_rpc(loop, active_rpc_timeout):
                @gen.coroutine
                def coro():
                    with dask.config.set(config):
                        s = False
                        for i in range(5):
                            try:
                                s, ws = yield start_cluster(
                                    ncores, scheduler, loop, security=security,
                                    Worker=Worker, scheduler_kwargs=scheduler_kwargs,
                                    worker_kwargs=worker_kwargs)
                            except Exception as e:
                                logger.error("Failed to start gen_cluster, retrying", exc_info=True)
                            else:
                                workers[:] = ws
                                args = [s] + workers
                                break
                        if s is False:
                            raise Exception("Could not start cluster")
                        if client:
                            c = yield Client(s.address, loop=loop, security=security,
                                             asynchronous=True, **client_kwargs)
                            args = [c] + args
                        try:
                            future = func(*args)
                            if timeout:
                                future = gen.with_timeout(timedelta(seconds=timeout),
                                                          future)
                            result = yield future
                            if s.validate:
                                s.validate_state()
                        finally:
                            if client:
                                yield c._close(fast=s.status == 'closed')
                            yield end_cluster(s, workers)
                            yield gen.with_timeout(timedelta(seconds=1),
                                                   cleanup_global_workers())

                        try:
                            c = yield default_client()
                        except ValueError:
                            pass
                        else:
                            yield c._close(fast=True)

                        raise gen.Return(result)

                result = loop.run_sync(coro, timeout=timeout * 2 if timeout else timeout)

            for w in workers:
                if getattr(w, 'data', None):
                    try:
                        w.data.clear()
                    except EnvironmentError:
                        # zict backends can fail if their storage directory
                        # was already removed
                        pass
                    del w.data
            DequeHandler.clear_all_instances()
            for w in _global_workers:
                w = w()
                w._close(report=False, executor_wait=False)
                if w.status == 'running':
                    w.close()
            del _global_workers[:]

        if PY3 and not WINDOWS and check_new_threads:
            start = time()
            while True:
                bad = [t for t, v in threading._active.items()
                       if t not in active_threads_start and
                      "Threaded" not in v.name and
                      "watch message" not in v.name and
                      "TCP-Executor" not in v.name]
                if not bad:
                    break
                else:
                    sleep(0.01)
                if time() > start + 5:
                    from distributed import profile
                    tid = bad[0]
                    thread = threading._active[tid]
                    call_stacks = profile.call_stack(sys._current_frames()[tid])
>                   assert False, (thread, call_stacks)
E                   AssertionError: (<Thread(Profile, started daemon 123146351648768)>, ['  File "/Users/hameerabbasi/anaconda/envs/dask/lib/python3.7/thr...asi/anaconda/envs/dask/lib/python3.7/site-packages/distributed/profile.py", line 251, in _watch
E                     \tsleep(interval)
E                     '])
E                   assert False

../../anaconda/envs/dask/lib/python3.7/site-packages/distributed/utils_test.py:945: AssertionError
----------------------------- Captured stderr call -----------------------------
distributed.scheduler - INFO - Clear task state
distributed.scheduler - INFO -   Scheduler at:     tcp://127.0.0.1:51130
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:51131
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:51131
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:51130
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          1
distributed.worker - INFO -                Memory:                   34.36 GB
distributed.worker - INFO -       Local Directory: /Users/hameerabbasi/Quansight/dask/dask-worker-space/worker-aohqtfy8
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -       Start worker at:      tcp://127.0.0.1:51132
distributed.worker - INFO -          Listening to:      tcp://127.0.0.1:51132
distributed.worker - INFO - Waiting to connect to:      tcp://127.0.0.1:51130
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -               Threads:                          2
distributed.worker - INFO -                Memory:                   34.36 GB
distributed.worker - INFO -       Local Directory: /Users/hameerabbasi/Quansight/dask/dask-worker-space/worker-uwr0s0vd
distributed.worker - INFO - -------------------------------------------------
distributed.scheduler - INFO - Register tcp://127.0.0.1:51131
distributed.scheduler - INFO - Register tcp://127.0.0.1:51132
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:51131
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Starting worker compute stream, tcp://127.0.0.1:51132
distributed.core - INFO - Starting established connection
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:51130
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO -         Registered to:      tcp://127.0.0.1:51130
distributed.worker - INFO - -------------------------------------------------
distributed.core - INFO - Starting established connection
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Receive client connection: Client-88a3d12e-5854-11e9-b2a9-e0d55eb2634b
distributed.core - INFO - Starting established connection
distributed.scheduler - INFO - Remove client Client-88a3d12e-5854-11e9-b2a9-e0d55eb2634b
distributed.scheduler - INFO - Remove client Client-88a3d12e-5854-11e9-b2a9-e0d55eb2634b
distributed.scheduler - INFO - Close client connection: Client-88a3d12e-5854-11e9-b2a9-e0d55eb2634b
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:51131
distributed.worker - INFO - Stopping worker at tcp://127.0.0.1:51132
distributed.scheduler - INFO - Scheduler closing...
distributed.scheduler - INFO - Scheduler closing all comms
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.worker - INFO - Connection to scheduler broken.  Reconnecting...
distributed.worker - INFO - -------------------------------------------------
distributed.worker - INFO - -------------------------------------------------
--------------------------- Captured stderr teardown ---------------------------
distributed.comm.tcp - WARNING - Closing dangling stream in <TCP local=tcp://127.0.0.1:51138 remote=tcp://127.0.0.1:51132>
```

**Comment by jrbourbeau** [comment]
It appears that the `gen_cluster` decorator from `distributed.utils_test` is raising the error here. What version of `distributed` is in your environment?

**Comment by hameerabbasi** [comment]
```
$ conda list | grep distributed
distributed               1.26.1                   pypi_0    pypi
```

File: dask/array/core.py [hunk 1]
```diff
@@ -3508,6 +3508,14 @@ def concatenate3(arrays):
 
     advanced = max(core.flatten(arrays, container=(list, tuple)),
                    key=lambda x: getattr(x, '__array_priority__', 0))
+
+    if not isinstance(advanced, np.ndarray) and hasattr(advanced, '__array_function__'):
```
Comment (mrocklin) [inline dask/array/core.py]: What if we have `__array_function__` but the NumPy version isn't advanced enough?

Comment (hameerabbasi) [reply dask/array/core.py]: I think some sort of way exists to detect if the protocol exists and is enabled. This is a flaw I spotted, but there's also the issue of the environment variable along with the version.

cc @shoyer is there an easy way to check for an (enabled) `__array_function__` protocol?

Comment (hameerabbasi) [reply dask/array/core.py]: We can probably add a check at import time.

**Comment by pentschev** [comment]
@hameerabbasi could you merge master into your branch? The codecov error should (hopefully) go away after https://github.com/dask/dask/pull/4671 was merged.

File: dask/array/core.py (hunk 1 above)
Comment (shoyer) [reply dask/array/core.py]: You could look at the NumPy version and environment variables. With 1.17 it will always be enabled. With 1.16 you need to opt in the with env variable.

Comment (hameerabbasi) [reply dask/array/core.py]: NEP-18 is still provisional, even in 1.17 right?

Comment (shoyer) [reply dask/array/core.py]: Yes, I suppose so.

Comment (hameerabbasi) [reply dask/array/core.py]: Then that route seems kinda flaky... @mrocklin If you want I could write a simple test to see if `__array_function__` exists on import and expose it. Maybe such a thing is useful in NumPy itself?

Comment (mrocklin) [reply dask/array/core.py]: Yes, I think that having a small utility function `is_nep18_active` or something similar would be useful here.  If we find it valuable in Dask then maybe that's good evidence to push upstream to NumPy.

Comment (hameerabbasi) [reply dask/array/core.py]: Opened a new PR (#4675) to that effect.

**Comment by jcrist** [comment]
cc @pentschev, @jrbourbeau - anything left to do here?

**Comment by pentschev** [comment]
@jcrist some concerns with this approach have been raised by @mrocklin in #4662, so I'll defer that to him.

**Comment by hameerabbasi** [comment]
@pentschev I believe this version should completely address @mrocklin's concerns.

File: dask/array/core.py [hunk 2]
```diff
@@ -16,6 +16,7 @@
 from threading import Lock
 import uuid
 import warnings
+from .utils import IS_NEP18_ACTIVE
```
Comment (jrbourbeau) [inline dask/array/core.py]: Moving this import inside the `concatenate3` function should fix the `ImportError`s [currently being raised](https://travis-ci.org/dask/dask/jobs/543666503#L1684)

**Comment by hameerabbasi** [comment]
Is anything else needed here? The linting error is unrelated to this PR.

**Comment by pentschev** [comment]
@hameerabbasi the lint error seems to be on https://github.com/dask/dask/pull/4669/files#diff-0dc21516366e88095fc3dbf2e9e62f9aR3515, it shows a different line number because it's merging upstream automatically. Also, there are [some errors](https://travis-ci.org/dask/dask/jobs/543839195#L1765) on the Python 3.7 build, which is the only build with NEP-18 enabled.

**Comment by pentschev** [comment]
I think the easiest (and perhaps safer) would be to check for `__array_function__` on all arrays, as @mrocklin suggested https://github.com/dask/dask/issues/4662#issuecomment-479185816, that alone may fix all issues.

File: dask/array/core.py [hunk 3]
```diff
@@ -3757,12 +3757,23 @@ def concatenate3(arrays):
...
+    from .utils import IS_NEP18_ACTIVE
+
     arrays = concrete(arrays)
     if not arrays:
         return np.empty(0)
 
     advanced = max(core.flatten(arrays, container=(list, tuple)),
                    key=lambda x: getattr(x, '__array_priority__', 0))
+
+    if if (IS_NEP18_ACTIVE and not all(np.ndarray.__array_function__ is
```
Comment (mrocklin) [inline dask/array/core.py]:
```suggestion
    if (IS_NEP18_ACTIVE and not all(np.ndarray.__array_function__ is
```

Comment (hameerabbasi) [reply dask/array/core.py]: Sorry, my laptop’s touchpad is way too sensitive and automatically pastes stuff.

File: dask/array/core.py [hunk 4]
```diff
@@ -3757,12 +3757,23 @@ def concatenate3(arrays):
...
+
     arrays = concrete(arrays)
     if not arrays:
         return np.empty(0)
 
     advanced = max(core.flatten(arrays, container=(list, tuple)),
                    key=lambda x: getattr(x, '__array_priority__', 0))
+
+    if (IS_NEP18_ACTIVE and not all(np.ndarray.__array_function__ is
+                                    getattr(arr, '__array_function__', np.ndarray.__array_function__) for arr in arrays)):
```
Comment (pentschev) [inline dask/array/core.py]:
```suggestion
                                    getattr(arr, '__array_function__', np.ndarray.__array_function__)
                                    for arr in arrays)):
```

Comment (pentschev) [reply dask/array/core.py]: @hameerabbasi sorry, looks like lint took the week off to pester you. This should fix it. :)

Comment (hameerabbasi) [reply dask/array/core.py]: Weird... I applied autopep8

**Comment by pentschev** [comment]
LGTM.

**Comment by pentschev** [comment]
@hameerabbasi thanks for the PR.

**Comment by hameerabbasi** [comment]
Anything else needed here?

**Comment by pentschev** [comment]
@mrocklin would you mind taking another look/merging?

File: dask/array/utils.py [hunk 5]
```diff
@@ -276,3 +276,4 @@ def __array_function__(self, *args, **kwargs):
 
 
 IS_NEP18_ACTIVE = _is_nep18_active()
+NDARRAY_ARRAY_FUNCTION = getattr(np.ndarray, '__array_function__', None)
```
Comment (mrocklin) [inline dask/array/utils.py]: Why this change?  Was the getattr call expensive in some way?  If not, I'd rather inline this and avoid hiding this.

**Comment by mrocklin** [comment]
Sorry for the delay here.  I added another small comment.

Also, is this worth testing?

**Comment by martindurant** [comment]
Ping to @hameerabbasi , hope you saw mrocklin's comments, above. Should take little time to get to to merged.

File: dask/array/utils.py (hunk 5 above)
Comment (hameerabbasi) [reply dask/array/utils.py]: Inlined. I didn't want there to be a `getattr` call in a loop. 😄

**Comment by mrocklin** [comment]
> hope you saw mrocklin's comments, above. Should take little time to get to to merged.

I think that the implementation is fine, though I'll admit to not fully understanding the motivation to avoid the getattr call (is the concern that this will have a significant impact on performance?  If so then I doubt it will be an issue)

I'm not sure I understand the proposed test, and it's unlikely that I'll have the time near-term to dive into it.

**Comment by hameerabbasi** [comment]
> I think that the implementation is fine, though I'll admit to not fully understanding the motivation to avoid the getattr call (is the concern that this will have a significant impact on performance? If so then I doubt it will be an issue)

Significant? I doubt it. It's two attribute lookups per item extra.

> I'm not sure I understand the proposed test, and it's unlikely that I'll have the time near-term to dive into it.

I'll add comments... What it basically does is "mock" an `ndarray` by encapsulating an `ndarray` and using protocols, and then tests whether various Dask functions work with this kind of array.

File: dask/array/tests/test_array_function.py [hunk 6]
```diff
@@ -104,3 +104,103 @@ def test_array_function_cupy_svd():
...
+
+        def __init__(self, arr):
+            self.arr = arr[...]
+
+        def __array__(self, **kwargs):
+            return np.asarray(self.arr, **kwargs)
+
+        def __array_function__(self, f, t, arrs, kw):
+            arrs = tuple(arr if not isinstance(arr, type(self)) else arr.arr for arr in arrs)
+            return type(self)(self.arr.__array_function__(f, (np.ndarray,), arrs, kw))
```
Comment (hameerabbasi) [inline dask/array/tests/test_array_function.py]: @shoyer @mhvk This "hack" might be interesting to you. ;) It'll need to be changed, but I couldn't find a better way to do it.

File: dask/array/core.py [hunk 7]
```diff
@@ -3757,12 +3757,27 @@ def concatenate3(arrays):
            [1, 2, 1, 2],
            [1, 2, 1, 2]])
     """
+    from .utils import IS_NEP18_ACTIVE
+    # We need this as __array_function__ may not exist on older NumPy versions.
+    # And to reduce verbosity.
+    NDARRAY_ARRAY_FUNCTION = getattr(np.ndarray, '__array_function__', None)
```
Comment (mhvk) [inline dask/array/core.py]: This is fragile: on numpy 1.16 it is defined, but dispatching is turned off. In astropy, I do:
```
if NUMPY_LT_1_17:  # pragma: no cover
    # pre 1.16, overrides.py did not exist; in 1.16, it existed, but
    # __array_function__ overrides were turned off by default.
    ARRAY_FUNCTION_ENABLED = (hasattr(np.core, 'overrides') and
                              np.core.overrides.ENABLE_ARRAY_FUNCTION)
else:
    # In 1.17, overrides are enabled by default, but it is still possible to
    # turn them off using an environment variable.  We use getattr since it
    # is planned to remove that possibility in later numpy versions.
    ARRAY_FUNCTION_ENABLED = getattr(np.core.overrides,
                                     'ENABLE_ARRAY_FUNCTION', True)
```

File: dask/array/tests/test_array_function.py (hunk 6 above)
Comment (mhvk) [reply dask/array/tests/test_array_function.py]: I realize it is for testing, but setting `types` to `(np.ndarray,)` will break functions in which another argument also overrides `__array_function__`. Instead, remove `type(self)` from `types`.
Also, if there are other types present, `ndarray.__array_function__(...)` should return `NotImplemented`, so you'd need to pass that on (I also check for `None` though I'm not sure that is relevant for functions - it is for `ufunc.at`).
Not sure if it helps, but the Quantity implementation is at https://github.com/astropy/astropy/blob/master/astropy/units/quantity.py#L1460 - note that I don't have to remove `Quantity` from the types since `Quantity` is an `ndarray` subclass.

File: dask/array/tests/test_array_function.py [hunk 8]
```diff
@@ -104,3 +104,103 @@ def test_array_function_cupy_svd():
...
+
+        __getitem__ = wrap('__getitem__')
+
+        __setitem__ = dispatch('__setitem__')
+
+        def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+            inputs = tuple(i if not isinstance(i, type(self)) else i.arr for i in inputs)
+            return type(self)(getattr(ufunc, method)(*inputs, **kwargs))
```
Comment (mhvk) [inline dask/array/tests/test_array_function.py]: As noted above, `ufunc.at` returns `None`.

File: dask/array/core.py (hunk 7 above)
Comment (hameerabbasi) [reply dask/array/core.py]: This isn't testing for `__array_function__`, that's in `IS_NEP18_ACTIVE`.

Comment (shoyer) [reply dask/array/core.py]: @mhvk That environment variables uses the wrong internal variable name. I think it's `ARRAY_FUNCTION_ENABLED` now on NumPy master (though I would suggest just checking the environment variable instead).

Comment (mhvk) [reply dask/array/core.py]: Duh, sorry for the noise.

**Comment by mrocklin** [comment]
@mhvk Any further comments here?  If the test here passes your standards then I'm quite happy to merge.

**Comment by mhvk** [comment]
@mrocklin - for tests, it really was fine as is, but now it is even better as it wouldn't lead users the wrong way (I find I often look at package tests to really understand what happens...)

**Comment by jakirkham** [comment]
Really great to see this in! 🎉 Thanks for working on this @hameerabbasi. 😄
